File: gesttravaux-symfony/src/Entity/Entrepreneur.php
<?php

namespace App\Entity;

use App\Repository\EntrepreneurRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;

/**
 * Entrepreneur entity
 *
 * Represents contractors who can bid on projects and perform work
 *
 * @author GestTravaux Pro
 */
#[ORM\Entity(repositoryClass: EntrepreneurRepository::class)]
#[ORM\Table(name: 'entrepreneurs')]
class Entrepreneur
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'idEntrepreneur')]
    private ?int $id = null;

    #[ORM\Column(length: 255, nullable: true)]
    private ?string $villeDeploiment = null;

    #[ORM\Column(length: 255)]
    private ?string $nomEntrepreneur = null;

    #[ORM\Column(length: 255)]
    private ?string $prenomEntrepreneur = null;

    #[ORM\Column(length: 255)]
    private ?string $emailEntrepreneur = null;

    #[ORM\Column(length: 45)]
    private ?string $telEntrepreneur = null;

    /**
     * @var Collection<int, Categorie>
     */
    #[ORM\ManyToMany(targetEntity: Categorie::class, inversedBy: 'entrepreneurs')]
    #[ORM\JoinTable(name: 'entrepreneurs_has_categories')]
    #[ORM\JoinColumn(name: 'entrepreneurs_idEntrepreneur', referencedColumnName: 'idEntrepreneur')]
    #[ORM\InverseJoinColumn(name: 'categories_idCategorie', referencedColumnName: 'idCategorie')]
    private Collection $categories;

    /**
     * @var Collection<int, Devis>
     */
    #[ORM\OneToMany(targetEntity: Devis::class, mappedBy: 'entrepreneur')]
    private Collection $devis;

    public function __construct()
    {
        $this->categories = new ArrayCollection();
        $this->devis = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getVilleDeploiment(): ?string
    {
        return $this->villeDeploiment;
    }

    public function setVilleDeploiment(?string $villeDeploiment): static
    {
        $this->villeDeploiment = $villeDeploiment;

        return $this;
    }

    public function getNomEntrepreneur(): ?string
    {
        return $this->nomEntrepreneur;
    }

    public function setNomEntrepreneur(string $nomEntrepreneur): static
    {
        $this->nomEntrepreneur = $nomEntrepreneur;

        return $this;
    }

    public function getPrenomEntrepreneur(): ?string
    {
        return $this->prenomEntrepreneur;
    }

    public function setPrenomEntrepreneur(string $prenomEntrepreneur): static
    {
        $this->prenomEntrepreneur = $prenomEntrepreneur;

        return $this;
    }

    public function getEmailEntrepreneur(): ?string
    {
        return $this->emailEntrepreneur;
    }

    public function setEmailEntrepreneur(string $emailEntrepreneur): static
    {
        $this->emailEntrepreneur = $emailEntrepreneur;

        return $this;
    }

    public function getTelEntrepreneur(): ?string
    {
        return $this->telEntrepreneur;
    }

    public function setTelEntrepreneur(string $telEntrepreneur): static
    {
        $this->telEntrepreneur = $telEntrepreneur;

        return $this;
    }

    /**
     * @return Collection<int, Categorie>
     */
    public function getCategories(): Collection
    {
        return $this->categories;
    }

    public function addCategory(Categorie $category): static
    {
        if (!$this->categories->contains($category)) {
            $this->categories->add($category);
        }

        return $this;
    }

    public function removeCategory(Categorie $category): static
    {
        $this->categories->removeElement($category);

        return $this;
    }

    /**
     * @return Collection<int, Devis>
     */
    public function getDevis(): Collection
    {
        return $this->devis;
    }

    public function addDevi(Devis $devi): static
    {
        if (!$this->devis->contains($devi)) {
            $this->devis->add($devi);
            $devi->setEntrepreneur($this);
        }

        return $this;
    }

    public function removeDevi(Devis $devi): static
    {
        if ($this->devis->removeElement($devi)) {
            // set the owning side to null (unless already changed)
            if ($devi->getEntrepreneur() === $this) {
                $devi->setEntrepreneur(null);
            }
        }

        return $this;
    }

    public function __toString(): string
    {
        return $this->prenomEntrepreneur . ' ' . $this->nomEntrepreneur;
    }
}

File: gesttravaux-java/src/main/java/com/immosync/gesttravaux/models/Entrepreneur.java
package com.immosync.gesttravaux.models;

import jakarta.persistence.*;
import java.util.HashSet;
import java.util.Set;

/**
 * Entity representing an entrepreneur who can work on construction projects.
 * Entrepreneurs can have multiple categories of expertise.
 *
 * @author GestTravaux Pro
 * @version 1.0
 */
@Entity
@Table(name = "ENTREPRENEURS")
public class Entrepreneur {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "idEntrepreneur")
    private Integer idEntrepreneur;

    @Column(name = "nomEntrepreneur", length = 100, nullable = false)
    private String nomEntrepreneur;

    @Column(name = "prenomEntrepreneur", length = 100, nullable = false)
    private String prenomEntrepreneur;

    @Column(name = "emailEntrepreneur", length = 100, nullable = false)
    private String emailEntrepreneur;

    @Column(name = "telEntrepreneur", length = 45)
    private String telEntrepreneur;

    @Column(name = "villeDeploiement", length = 255)
    private String villeDeploiement;

    @ManyToMany(fetch = FetchType.EAGER, cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(
        name = "ENTREPRENEURS_has_CATEGORIES",
        joinColumns = @JoinColumn(name = "ENTREPRENEURS_idEntrepreneur"),
        inverseJoinColumns = @JoinColumn(name = "CATEGORIES_idCategorie")
    )
    private Set<Categorie> categories = new HashSet<>();

    @OneToMany(mappedBy = "entrepreneur", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<Devis> devisList = new HashSet<>();

    /**
     * Default constructor required by JPA.
     */
    public Entrepreneur() {
    }

    /**
     * Constructor with essential fields.
     *
     * @param nomEntrepreneur entrepreneur last name
     * @param prenomEntrepreneur entrepreneur first name
     * @param emailEntrepreneur entrepreneur email
     */
    public Entrepreneur(String nomEntrepreneur, String prenomEntrepreneur, String emailEntrepreneur) {
        this.nomEntrepreneur = nomEntrepreneur;
        this.prenomEntrepreneur = prenomEntrepreneur;
        this.emailEntrepreneur = emailEntrepreneur;
    }

    // Getters and Setters

    /**
     * Gets the entrepreneur ID.
     *
     * @return the entrepreneur ID
     */
    public Integer getIdEntrepreneur() {
        return idEntrepreneur;
    }

    /**
     * Sets the entrepreneur ID.
     *
     * @param idEntrepreneur the entrepreneur ID
     */
    public void setIdEntrepreneur(Integer idEntrepreneur) {
        this.idEntrepreneur = idEntrepreneur;
    }

    /**
     * Gets the entrepreneur last name.
     *
     * @return the last name
     */
    public String getNomEntrepreneur() {
        return nomEntrepreneur;
    }

    /**
     * Sets the entrepreneur last name.
     *
     * @param nomEntrepreneur the last name
     */
    public void setNomEntrepreneur(String nomEntrepreneur) {
        this.nomEntrepreneur = nomEntrepreneur;
    }

    /**
     * Gets the entrepreneur first name.
     *
     * @return the first name
     */
    public String getPrenomEntrepreneur() {
        return prenomEntrepreneur;
    }

    /**
     * Sets the entrepreneur first name.
     *
     * @param prenomEntrepreneur the first name
     */
    public void setPrenomEntrepreneur(String prenomEntrepreneur) {
        this.prenomEntrepreneur = prenomEntrepreneur;
    }

    /**
     * Gets the entrepreneur email.
     *
     * @return the email
     */
    public String getEmailEntrepreneur() {
        return emailEntrepreneur;
    }

    /**
     * Sets the entrepreneur email.
     *
     * @param emailEntrepreneur the email
     */
    public void setEmailEntrepreneur(String emailEntrepreneur) {
        this.emailEntrepreneur = emailEntrepreneur;
    }

    /**
     * Gets the entrepreneur phone number.
     *
     * @return the phone number
     */
    public String getTelEntrepreneur() {
        return telEntrepreneur;
    }

    /**
     * Sets the entrepreneur phone number.
     *
     * @param telEntrepreneur the phone number
     */
    public void setTelEntrepreneur(String telEntrepreneur) {
        this.telEntrepreneur = telEntrepreneur;
    }

    /**
     * Gets the deployment city.
     *
     * @return the deployment city
     */
    public String getVilleDeploiement() {
        return villeDeploiement;
    }

    /**
     * Sets the deployment city.
     *
     * @param villeDeploiement the deployment city
     */
    public void setVilleDeploiement(String villeDeploiement) {
        this.villeDeploiement = villeDeploiement;
    }

    /**
     * Gets the categories associated with this entrepreneur.
     *
     * @return set of categories
     */
    public Set<Categorie> getCategories() {
        return categories;
    }

    /**
     * Sets the categories associated with this entrepreneur.
     *
     * @param categories set of categories
     */
    public void setCategories(Set<Categorie> categories) {
        this.categories = categories;
    }

    /**
     * Gets the list of devis for this entrepreneur.
     *
     * @return set of devis
     */
    public Set<Devis> getDevisList() {
        return devisList;
    }

    /**
     * Sets the list of devis for this entrepreneur.
     *
     * @param devisList set of devis
     */
    public void setDevisList(Set<Devis> devisList) {
        this.devisList = devisList;
    }

    /**
     * Adds a category to this entrepreneur.
     *
     * @param categorie the category to add
     */
    public void addCategorie(Categorie categorie) {
        categories.add(categorie);
        categorie.getEntrepreneurs().add(this);
    }

    /**
     * Removes a category from this entrepreneur.
     *
     * @param categorie the category to remove
     */
    public void removeCategorie(Categorie categorie) {
        categories.remove(categorie);
        categorie.getEntrepreneurs().remove(this);
    }

    /**
     * Gets the full name of the entrepreneur.
     *
     * @return full name (first name + last name)
     */
    public String getFullName() {
        return prenomEntrepreneur + " " + nomEntrepreneur;
    }

    @Override
    public String toString() {
        return getFullName();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Entrepreneur)) return false;
        Entrepreneur that = (Entrepreneur) o;
        return idEntrepreneur != null && idEntrepreneur.equals(that.getIdEntrepreneur());
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}
